### PR TITLE
[consensus] Replace `Epocher` with `Monitor`

### PR DESCRIPTION
**Pull request by patrick-ogrady**
## Changes
* Introduce `Monitor` trait to replace one-off `Epocher` (which caused some confusion with devs)
* Apply stricter verification rules in `mocks`
* Fix handling of parent threshold signatures in `prover`
* Adjust pruning rules to ensure we don't send a `Finalized` message for the same height twice

### CODE REVIEW

File: consensus/src/ordered_broadcast/engine.rs [hunk 1]
```diff
@@ -389,17 +392,26 @@ impl<
 
                     // Handle the parent threshold signature
                     if let Some(parent) = node.parent.as_ref() {
-                        self.handle_threshold(&node.chunk, parent.epoch, parent.threshold).await;
+                        let Some(parent_height) = node.chunk.height.checked_sub(1) else {
+                            warn!(?sender, height=node.chunk.height, "parent height underflow");
+                            continue;
+                        };
+                        let parent_chunk = parsed::Chunk {
```
Comment (patrick-ogrady) [inline consensus/src/ordered_broadcast/engine.rs:399]: This was a bug

File: consensus/src/ordered_broadcast/ingress.rs
Comment (patrick-ogrady) [inline consensus/src/ordered_broadcast/ingress.rs:1]: After the refactor, this file was no longer used.

File: consensus/src/ordered_broadcast/mocks/committer.rs [hunk 2]
```diff
@@ -53,19 +52,15 @@ impl<C: Scheme, D: Digest> Committer<C, D> {
...
                     // The prover checks the validity of the threshold signature when deserializing
-                    let prover = Prover::<C, D>::new(&self.namespace, self.public);
-                    let context = match prover.deserialize_threshold(proof) {
-                        Some((context, _payload, _epoch, _threshold)) => context,
-                        None => {
-                            error!("invalid proof");
-                            continue;
-                        }
-                    };
+                    let (context, _, _, _) =
```
Comment (patrick-ogrady) [inline consensus/src/ordered_broadcast/mocks/committer.rs]: We now enforce proofs MUST be properly verified.

File: consensus/src/ordered_broadcast/engine.rs [hunk 3]
```diff
@@ -389,17 +392,26 @@ impl<
 
                     // Handle the parent threshold signature
                     if let Some(parent) = node.parent.as_ref() {
-                        self.handle_threshold(&node.chunk, parent.epoch, parent.threshold).await;
```
Comment (patrick-ogrady) [inline consensus/src/ordered_broadcast/engine.rs:392]: Using `&node.chunk` here is a bug

File: consensus/src/ordered_broadcast/engine.rs (hunk 1 above)
Comment (patrick-ogrady) [inline consensus/src/ordered_broadcast/engine.rs]: This is generated inside of `validate_node` and we should reuse that.

File: consensus/src/ordered_broadcast/engine.rs [hunk 4]
```diff
@@ -547,6 +565,8 @@ impl<
         threshold: group::Signature,
     ) {
         // Set the threshold signature, returning early if it already exists
+        //
+        // TODO: this prevents us from ever sending a finalization message recovered from a partial signature
```
Comment (patrick-ogrady) [inline consensus/src/ordered_broadcast/engine.rs:569]: We need to refactor the handling of `handle_threshold` to allow messages to be sent when a partial signature is recovered.

File: consensus/src/ordered_broadcast/ack_manager.rs [hunk 5]
```diff
@@ -134,9 +134,14 @@ impl<D: Digest, P: Array> AckManager<D, P> {
             return false;
         }
 
-        // Prune lower-height for this sequencer
+        // Prune all entries with height less than the parent
+        //
+        // This approach ensures we don't accidentally notify the application of a threshold signature multiple
```
Comment (patrick-ogrady) [inline consensus/src/ordered_broadcast/ack_manager.rs:139]: This is another edge case we now test for.

File: consensus/src/ordered_broadcast/engine.rs [hunk 6]
```diff
@@ -379,27 +381,34 @@ impl<
...
+                    if let Some(parent_chunk) = result {
+                        let parent = node.parent.as_ref().unwrap();
+                        self.handle_threshold(&parent_chunk, parent.epoch, parent.threshold).await;
                     }
 
-                    // Process the new node
+                    // Process the node
+                    //
+                    // Note, this node may be a duplicate. If it is, we will attempt to verify it and vote
+                    // on it again (our original vote may have been lost).
```
Comment (patrick-ogrady) [inline consensus/src/ordered_broadcast/engine.rs:397]: TODO: avoid duplicate verification of resent nodes

Comment (patrick-ogrady) [reply consensus/src/ordered_broadcast/engine.rs:397]: https://github.com/commonwarexyz/monorepo/issues/650

File: consensus/src/lib.rs [hunk 7]
```diff
@@ -174,5 +174,23 @@ cfg_if::cfg_if! {
...
         }
+
+        /// Monitor is the interface an external actor can use to observe the progress of a consensus implementation.
+        ///
+        /// Monitor is used to implement mechanisms that share the same set of active participants as consensus and/or
+        /// perform some activity that requires some synchronization with the progress of consensus.
+        ///
+        /// Monitor can be implemented using [`Committer`](crate::Committer) to avoid introducing complexity
+        /// into any particular consensus implementation.
+        pub trait Monitor: Clone + Send + 'static {
```
Comment (patrick-ogrady) [inline consensus/src/lib.rs:185]: I considered just having `ordered_broadcast` implement `Committer` (and use that) but I figured there may be multiple consumers of that stream (and/or some implementation may prefer to expose different types of monitors directly). Open to feedback on this decision but think it is at least an improvement over `Epocher`.

File: consensus/src/ordered_broadcast/engine.rs [hunk 8]
```diff
@@ -325,15 +318,26 @@ impl<
                 },
 
                 // Handle refresh epoch deadline
-                _ = refresh_epoch => {
-                    debug!("refresh epoch");
-                    // Simply continue; the epoch will be refreshed on the next iteration.
+                epoch = epoch_updates.next() => {
+                    // Error handling
+                    let epoch = match epoch{
```
Comment (BrendanChou) [inline consensus/src/ordered_broadcast/engine.rs]: Is this linted? I usually feel like there would be a space between `epoch{`

**Review by BrendanChou — COMMENTED** [review]
`ordered_broadcast/engine.rs` and `ordered_broadcast/config.rs` look good to me

File: consensus/src/ordered_broadcast/engine.rs (hunk 8 above)
Comment (patrick-ogrady) [reply consensus/src/ordered_broadcast/engine.rs]: I think these macros are really messing with the linter...

File: consensus/src/lib.rs [hunk 9]
```diff
@@ -174,5 +174,23 @@ cfg_if::cfg_if! {
...
+            /// Index is the type used to indicate the in-progress consensus decision.
+            type Index;
+
+            /// Latest index known by the consensus implementation.
+            fn latest(&self) -> Self::Index;
+
+            /// Create a channel that will receive updates when the latest index changes.
+            fn subscribe(&mut self) -> mpsc::Receiver<Self::Index>;
```
Comment (patrick-ogrady) [inline consensus/src/lib.rs:193]: This removes the need to poll for updates.

File: consensus/src/ordered_broadcast/mod.rs [hunk 10]
```diff
@@ -577,12 +597,76 @@ mod tests {
...
+            let link = Link {
+                latency: 10.0,
+                jitter: 1.0,
+                success_rate: 1.0,
+            };
+            link_validators(&mut oracle, &pks, Action::Link(link), None).await;
+            await_committers(
+                context.with_label("committer"),
+                &committers,
+                (max_height + 100, 112),
```
Comment (patrick-ogrady) [inline consensus/src/ordered_broadcast/mod.rs:667]: I modified `await_committers` so that we can assert the epoch of the latest confirmed chunk has increased (not that there was an issue).

File: consensus/src/ordered_broadcast/engine.rs [hunk 11]
```diff
@@ -325,15 +318,26 @@ impl<
...
-                _ = refresh_epoch => {
-                    debug!("refresh epoch");
-                    // Simply continue; the epoch will be refreshed on the next iteration.
+                epoch = epoch_updates.next() => {
+                    // Error handling
+                    let epoch = match epoch {
+                        Some(epoch) => epoch,
+                        None => {
+                            error!("epoch subscription failed");
+                            break;
```
[the remark this answers is not in the document]
Comment (patrick-ogrady) [reply consensus/src/ordered_broadcast/engine.rs]: This should only happen if the monitor has shutdown.

File: consensus/src/ordered_broadcast/engine.rs [hunk 12]
```diff
@@ -459,6 +455,12 @@ impl<
                 },
             }
         }
+
+        // Close all journals, regardless of how we exit the loop
```
Comment (patrick-ogrady) [inline consensus/src/ordered_broadcast/engine.rs:459]: The closing of channels we rely on could have caused us to exit without closing the journals. This wouldn't create a correctness issue (given how we use `sync()`) but would prevent graceful handling.

File: consensus/src/lib.rs [hunk 13]
```diff
@@ -150,7 +150,7 @@ cfg_if::cfg_if! {
         ///
         /// ## Synchronization
         ///
-        /// The same considerations for `Supervisor` apply here.
+        /// The same considerations for [`Supervisor`](crate::Supervisor) apply here.
```
Comment (BrendanChou) [inline consensus/src/lib.rs:153]: nit: you don't need `(crate::Supervisor)` here since `Supervisor` is in scope, the rustdoc can find it automatically if you just use ```[`Supervisor`]```

File: consensus/src/lib.rs (hunk 9 above)
Comment (BrendanChou) [inline consensus/src/lib.rs:193]: To avoid a race condition during async, should `subscribe` simply return a tuple `(latest, receiver)`?

Otherwise you could either:
- Call `latest` then `subscribe`, possibly dropping an update that occurs between the two calls
- Call `subscribe` then `latest`, possibly processing a message from the subscription that updates to a value less than the `latest` value

I think (?) this can happen in a concurrent environment even though the two functions are not `async`

Comment (patrick-ogrady) [reply consensus/src/lib.rs:193]: Ah interesting. Good catch.
